File: sorting-searching/sorting-algorithms/Branchless-sorting/branchless-sorting.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void branchlessCompareSwap(int &a, int &b) {
    int minValue = min(a, b);
    int maxValue = max(a, b);

    a = minValue;
    b = maxValue;
}

void branchlessSort(vector<int>& arr) {
    int n = arr.size();

    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            branchlessCompareSwap(arr[j], arr[j + 1]);
        }
    }
}

void printArray(const vector<int>& arr) {
    for (int value : arr) {
        cout << value << " ";
    }
    cout << endl;
}

int main() {
    vector<int> arr = {8, 3, 7, 4, 9, 2, 6, 1};

    cout << "Original array: ";
    printArray(arr);

    branchlessSort(arr);

    cout << "Sorted array: ";
    printArray(arr);

    return 0;
}

File: sorting-searching/sorting-algorithms/Branchless-sorting/BranchlessSorting.java
import java.util.Arrays;

public class BranchlessSorting {

    static int[] branchlessCompareSwap(int a, int b) {
        int minValue = Math.min(a, b);
        int maxValue = Math.max(a, b);

        return new int[]{minValue, maxValue};
    }

    static void branchlessSort(int[] arr) {
        int n = arr.length;

        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                int[] values = branchlessCompareSwap(arr[j], arr[j + 1]);

                arr[j] = values[0];
                arr[j + 1] = values[1];
            }
        }
    }

    public static void main(String[] args) {
        int[] arr = {8, 3, 7, 4, 9, 2, 6, 1};

        System.out.println("Original array: " + Arrays.toString(arr));

        branchlessSort(arr);

        System.out.println("Sorted array: " + Arrays.toString(arr));
    }
}
```
